File: go-sidecar/signer/vc.go
// W3C Verifiable Credential envelope for Vouch Protocol.
//
// Mirrors vouch/vc.py and typescript/src/vc.ts. Builds a VouchCredential
// per W3C CG Report §5: a W3C VC Data Model 2.0 credential carrying an
// agent's intent, optional reputation, and optional delegation chain,
// secured by a Data Integrity proof (eddsa-jcs-2022).

package signer

import (
	"crypto/rand"
	"errors"
	"fmt"
	"time"
)

const (
	VCContextV2          = "https://www.w3.org/ns/credentials/v2"
	VouchContextV1       = "https://vouch-protocol.com/contexts/v1"
	VCType               = "VerifiableCredential"
	VouchCredentialType  = "VouchCredential"
	SessionVoucherType   = "SessionVoucher"
	ProtocolVersion      = "1.0"
)

// Intent describes the action being authorized. action, target, and resource
// are REQUIRED per W3C CG Report §5.4.1.
type Intent struct {
	Action   string         `json:"action"`
	Target   string         `json:"target"`
	Resource string         `json:"resource"`
	Extra    map[string]any `json:"-"`
}

// DelegationLink is one link in a delegation chain.
type DelegationLink struct {
	Issuer            string         `json:"issuer"`
	Subject           string         `json:"subject"`
	Intent            map[string]any `json:"intent"`
	ValidFrom         string         `json:"validFrom,omitempty"`
	ValidUntil        string         `json:"validUntil,omitempty"`
	ParentProofValue  string         `json:"parentProofValue,omitempty"`
}

// BuildVouchCredentialOptions configures BuildVouchCredential.
type BuildVouchCredentialOptions struct {
	IssuerDID        string
	Intent           map[string]any
	ValidSeconds     int
	ReputationScore  *int
	DelegationChain  []map[string]any
	CredentialID     string
	ValidFrom        time.Time

	// CredentialStatus is an optional W3C credentialStatus entry, typically
	// built via BuildStatusListEntry to reference a BitstringStatusListCredential
	// (W3C CG Report §11.2). When non-nil, it is attached to the credential as
	// the `credentialStatus` property.
	CredentialStatus map[string]any
}

// BuildVouchCredential constructs an unsigned Vouch Credential. The caller
// attaches a Data Integrity proof via BuildDataIntegrityProof.
//
// Returns the credential as a map[string]any so it composes naturally with
// the JCS canonicalizer.
func BuildVouchCredential(opts BuildVouchCredentialOptions) (map[string]any, error) {
	if err := validateIntent(opts.Intent); err != nil {
		return nil, err
	}

	issuedAt := opts.ValidFrom
	if issuedAt.IsZero() {
		issuedAt = time.Now().UTC()
	} else {
		issuedAt = issuedAt.UTC()
	}

	validSeconds := opts.ValidSeconds
	if validSeconds <= 0 {
		validSeconds = 300
	}
	expiresAt := issuedAt.Add(time.Duration(validSeconds) * time.Second)

	subject := map[string]any{
		"id":           opts.IssuerDID,
		"vouchVersion": ProtocolVersion,
		"intent":       opts.Intent,
	}

	if opts.ReputationScore != nil {
		score := *opts.ReputationScore
		if score < 0 {
			score = 0
		}
		if score > 100 {
			score = 100
		}
		subject["reputationScore"] = score
	}

	if len(opts.DelegationChain) > 0 {
		// Convert []map[string]any to []any for JSON shape.
		chain := make([]any, len(opts.DelegationChain))
		for i, link := range opts.DelegationChain {
			chain[i] = link
		}
		subject["delegationChain"] = chain
	}

	credID := opts.CredentialID
	if credID == "" {
		uuid, err := newUUIDURN()
		if err != nil {
			return nil, err
		}
		credID = uuid
	}

	vc := map[string]any{
		"@context":          []any{VCContextV2, VouchContextV1},
		"id":                credID,
		"type":              []any{VCType, VouchCredentialType},
		"issuer":            opts.IssuerDID,
		"validFrom":         formatISO8601(issuedAt),
		"validUntil":        formatISO8601(expiresAt),
		"credentialSubject": subject,
	}

	if opts.CredentialStatus != nil {
		vc["credentialStatus"] = opts.CredentialStatus
	}

	return vc, nil
}

func validateIntent(intent map[string]any) error {
	if intent == nil {
		return errors.New("intent must not be nil")
	}
	for _, required := range []string{"action", "target", "resource"} {
		v, ok := intent[required]
		if !ok || v == nil {
			return fmt.Errorf(
				"intent.%s is REQUIRED (W3C CG Report §5.4.1), Vouch credentials MUST bind to a concrete resource",
				required,
			)
		}
		if s, isStr := v.(string); isStr && s == "" {
			return fmt.Errorf(
				"intent.%s is REQUIRED (W3C CG Report §5.4.1), Vouch credentials MUST bind to a concrete resource",
				required,
			)
		}
	}
	return nil
}

// newUUIDURN returns a "urn:uuid:" prefixed UUID v4 string.
func newUUIDURN() (string, error) {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", fmt.Errorf("uuid generation: %w", err)
	}
	// Set version (4) and variant (10).
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf(
		"urn:uuid:%08x-%04x-%04x-%04x-%012x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:16],
	), nil
}

File: packages/sdk-ts/src/vc.ts
/**
 * Verifiable Credential envelope for Vouch Protocol.
 *
 * Mirrors `vouch/vc.py`. Builds a `VouchCredential` per Specification §5:
 * a VC Data Model 2.0 credential carrying an agent's intent, optional
 * reputation, and optional delegation chain, secured by a Data Integrity
 * proof (eddsa-jcs-2022).
 */

export const VC_CONTEXT_V2 = 'https://www.w3.org/ns/credentials/v2';
export const VOUCH_CONTEXT_V1 = 'https://vouch-protocol.com/contexts/v1';

export const VC_TYPE = 'VerifiableCredential';
export const VOUCH_CREDENTIAL_TYPE = 'VouchCredential';
export const SESSION_VOUCHER_TYPE = 'SessionVoucher';

export const PROTOCOL_VERSION = '1.0';

export interface Intent {
  action: string;
  target: string;
  resource: string;
  [extra: string]: unknown;
}

export interface DelegationLink {
  issuer: string;
  subject: string;
  intent: Intent;
  validFrom?: string;
  validUntil?: string;
  parentProofValue?: string;
}

export interface CredentialSubject {
  id: string;
  vouchVersion: string;
  intent: Intent;
  reputationScore?: number;
  delegationChain?: DelegationLink[];
}

export interface VouchCredential {
  '@context': string[];
  id: string;
  type: string[];
  issuer: string;
  validFrom: string;
  validUntil: string;
  credentialSubject: CredentialSubject;
  credentialStatus?: unknown;
  proof?: unknown;
}

export interface BuildVouchCredentialOptions {
  issuerDid: string;
  intent: Intent;
  validSeconds?: number;
  reputationScore?: number;
  delegationChain?: DelegationLink[];
  credentialId?: string;
  validFrom?: Date;
  credentialStatus?: Record<string, unknown>;
}

/**
 * Construct an unsigned Vouch Credential. Caller is responsible for
 * attaching a Data Integrity proof via `buildProof`.
 *
 * `credentialStatus` is typically built via `buildStatusListEntry` to
 * reference a BitstringStatusListCredential (Specification §11.2).
 */
export function buildVouchCredential(
  opts: BuildVouchCredentialOptions
): VouchCredential {
  validateIntent(opts.intent);

  const issuedAt = opts.validFrom ?? new Date();
  const validSeconds = opts.validSeconds ?? 300;
  const expiresAt = new Date(issuedAt.getTime() + validSeconds * 1000);

  const subject: CredentialSubject = {
    id: opts.issuerDid,
    vouchVersion: PROTOCOL_VERSION,
    intent: opts.intent,
  };

  if (opts.reputationScore !== undefined) {
    subject.reputationScore = Math.max(
      0,
      Math.min(100, Math.floor(opts.reputationScore))
    );
  }

  if (opts.delegationChain && opts.delegationChain.length > 0) {
    subject.delegationChain = opts.delegationChain;
  }

  const vc: VouchCredential = {
    '@context': [VC_CONTEXT_V2, VOUCH_CONTEXT_V1],
    id: opts.credentialId ?? newUuidUrn(),
    type: [VC_TYPE, VOUCH_CREDENTIAL_TYPE],
    issuer: opts.issuerDid,
    validFrom: iso(issuedAt),
    validUntil: iso(expiresAt),
    credentialSubject: subject,
  };

  if (opts.credentialStatus !== undefined) {
    vc.credentialStatus = opts.credentialStatus;
  }

  return vc;
}

function validateIntent(intent: Intent): void {
  if (!intent || typeof intent !== 'object') {
    throw new TypeError('intent must be an object');
  }
  for (const required of ['action', 'target', 'resource'] as const) {
    const v = intent[required];
    if (v === undefined || v === null || v === '') {
      throw new Error(
        `intent.${required} is REQUIRED (Specification §5.4.1), ` +
        'Vouch credentials MUST bind to a concrete resource'
      );
    }
  }
}

function newUuidUrn(): string {
  // Resolve a UUID v4 from Web Crypto so this module stays free of the Node
  // `crypto` import and is importable in React Native (Hermes). Works in:
  //   - Node 18+ and modern browsers (globalThis.crypto.randomUUID)
  //   - React Native with a getRandomValues polyfill but no randomUUID
  //     (e.g. react-native-get-random-values) via the v4 fallback below.
  const webCrypto: Crypto | undefined = (globalThis as { crypto?: Crypto }).crypto;
  if (webCrypto && typeof webCrypto.randomUUID === 'function') {
    return `urn:uuid:${webCrypto.randomUUID()}`;
  }
  if (webCrypto && typeof webCrypto.getRandomValues === 'function') {
    return `urn:uuid:${uuidV4FromRandomValues(webCrypto)}`;
  }
  throw new Error(
    'No Web Crypto RNG available to generate a credential id. Pass ' +
    '`credentialId` explicitly, or install a getRandomValues polyfill ' +
    '(e.g. react-native-get-random-values).'
  );
}

function uuidV4FromRandomValues(webCrypto: Crypto): string {
  const bytes = new Uint8Array(16);
  webCrypto.getRandomValues(bytes);
  bytes[6] = (bytes[6] & 0x0f) | 0x40; // version 4
  bytes[8] = (bytes[8] & 0x3f) | 0x80; // variant 10xx
  const hex: string[] = [];
  for (let i = 0; i < 16; i++) hex.push(bytes[i].toString(16).padStart(2, '0'));
  return (
    hex[0] + hex[1] + hex[2] + hex[3] + '-' +
    hex[4] + hex[5] + '-' +
    hex[6] + hex[7] + '-' +
    hex[8] + hex[9] + '-' +
    hex[10] + hex[11] + hex[12] + hex[13] + hex[14] + hex[15]
  );
}

function iso(d: Date): string {
  return d.toISOString().replace(/\.\d{3}Z$/, 'Z');
}
